### PR TITLE
Scan core `civicrm/managed` directory for entity declarations

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Allows managed entities to be shipped with CiviCRM core, independent of any extensions.

Before
----------------------------------------
Managed entities can only be packaged in extensions.

After
----------------------------------------
They can also be packaged in CiviCRM core.

Technical Details
----------------------------------------
Includes 'CiviCRM Core' as an option for 'module' wherever managed entities are displayed.

### CODE REVIEW

File: CRM/Core/BAO/Managed.php [hunk 1]
```diff
@@ -54,4 +79,12 @@ public static function isApi4ManagedType(string $entityName) {
     return $type && in_array('ManagedEntity', $type, TRUE);
   }
 
+  /**
+   * Options callback for `base_module`.
+   * @return array
+   */
+  public static function getBaseModules(): array {
+    return array_merge(['civicrm' => ts('CiviCRM Core')], CRM_Core_PseudoConstant::getExtensions());
```
Comment (totten) [inline CRM/Core/BAO/Managed.php:87]: @colemanw -- Ah, let's see if I follow. When `civicrm_managed` was exposed to APIv4, it needed a method to lookup legal values of `module`. `getExtensions()`  was used. But this use-case shows that it's not quite accurate enough for this purpose. So this patch adds a method which should be more accurate.

There is a pre-existing method that is *even more* accurate -- `CRM_Core_Module::getAll()`. That's the method used by  [`CRM_Core_ManagedEntities`](https://github.com/civicrm/civicrm-core/blob/6658626809ee1837cced2413509996fb23d2c13a/CRM/Core/ManagedEntities.php#L39) and corresponds to some other options in [`hook_managed` docs](https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_managed/) (e.g. `drupal.mymodule`). Specifically, `Module::getAll()` provides:

* `civicrm`
* CiviCRM module-extensions (full `key`)
* Drupal/Backdrop modules (with `drupal.*` prefix)
* Joomla plugin-extensions (with `joomla.*` prefix)

(The essence of a "managed" entity is to install/delete based on whether the `module` is active/inactive, which requires inspecting `module` statuses. `CRM_Core_Module`  is the list of inspectable modules.)

Comment (colemanw) [reply CRM/Core/BAO/Managed.php:87]: @totten `CRM_Core_Module::getAll()` doesn't return human-readable labels.

File: distmaker/dists/common.sh [hunk 2]
```diff
@@ -82,7 +82,7 @@ function dm_install_core() {
   local repo="$1"
   local to="$2"
 
-  for dir in ang css i js PEAR templates bin CRM api extern Reports install mixin settings Civi partials release-notes xml setup sql/civicrm_data ; do
+  for dir in ang css i js PEAR templates bin CRM api extern Reports install managed mixin settings Civi partials release-notes xml setup sql/civicrm_data ; do
```
Comment (totten) [inline distmaker/dists/common.sh:85]: Nice. :+1:  That one is easy to forget.

File: CRM/Core/BAO/Managed.php (hunk 1 above)
Comment (colemanw) [reply CRM/Core/BAO/Managed.php:87]: @totten the user-facing label is shown for example in the SK UI and in the Afform UI, both in the "Package" column and in the "Filter by Package" dropdown:
image

File: distmaker/dists/common.sh (hunk 2 above)
Comment (colemanw) [reply distmaker/dists/common.sh:85]: @totten I looked for other places where we hard-code root directory names but didn't see any more. Unless you know of some.

File: CRM/Core/BAO/Managed.php (hunk 1 above)
Comment (totten) [reply CRM/Core/BAO/Managed.php:87]: Aah, cool. I see it now. :+1:

File: CRM/Core/BAO/Managed.php [hunk 3]
```diff
@@ -20,6 +20,31 @@
  */
 class CRM_Core_BAO_Managed extends CRM_Core_DAO_Managed implements Civi\Core\HookInterface {
 
+  /**
+   * Scan core `civicrm/managed` directory for entity declarations.
+   *
+   * Note: This is similar to the `mgd-php` mixin for extensions, but slightly stricter:
+   *  - It doesn't scan any directory outside `managed/`
+   *  - It doesn't allow omitting `$params['version']`
+   * TODO: Consider making a 2.0 version of the extension mixin using this code, for consistent strictness.
```
Comment (totten) [inline CRM/Core/BAO/Managed.php:29]: That proposal probably merits a Gitlab issue/discussion.

**Pull request merged**